base/contextvariables.jl

```diff
@@ -0,0 +1,415 @@
+`ctx` where `k => v` is inserted to `ctx′` for each `k => Some(v)` in `kvs`
+and `k` is deleted from `ctx′` for each `k => nothing` in `kvs`.
+"""
+function merge_ctxvars(ctx, kvs)
+    # Assumption: eltype(kvs) <: Pair{UUID,<:Union{Some,Nothing}}
+    if isempty(kvs)
+        return ctx
+    else
+        # Copy-or-create-on-write:
+        vars = ctx === nothing ? Dict{UUID,Any}() : copy(ctx)
+        for (k, v) in kvs
+            if v === nothing
+                delete!(vars, k)
+            else
+                vars[k] = something(v)
+            end
+        end
+        isempty(vars) && return nothing  # should we?
+        return vars
+    end
+end
+
+get_task_ctxvars(t = current_task()) = t.ctxvars
+set_task_ctxvars(t, ctx) = t.ctxvars = ctx
+set_task_ctxvars(ctx) = set_task_ctxvars(current_task(), ctx)
+
+struct _NoValue end
+
+"""
+    get(var::ContextVar{T}) -> Union{Some{T},Nothing}
+
+Return `Some(value)` if `value` is assigned to `var`.  Return `nothing` if
+unassigned.
+"""
+function get(var::ContextVar{T}) where {T}
+    ctx = get_task_ctxvars()
+    if ctx === nothing
+        var.has_default && return Some(var.default)
+        return nothing
+    end
+    if var.has_default
+        return Some(get(ctx, var.key, var.default)::T)
+    else
+        y = get(ctx, var.key, _NoValue())
+        y isa _NoValue || return Some(ctx[var.key]::T)
+    end
+    return nothing
+end
+
+"""
+    getindex(var::ContextVar{T}) -> value::T
+
+Return the `value` assigned to `var`.  Throw a `KeyError` if unassigned.
+"""
+function getindex(var::ContextVar{T}) where {T}
+    maybe = get(var)
+    maybe === nothing && throw(KeyError(var))
+    return something(maybe)::T
+end
+
+setindex!(var::ContextVar, value) = set!(var, Some(value))
+delete!(var::ContextVar) = set!(var, nothing)
+
+"""
+    set!(var::ContextVar, Some(value))
+    set!(var::ContextVar, nothing)
+
+Set the value of context variable `var` to `value` or delete it.
+"""
+function set!(var::ContextVar{T}, value::Union{Some,Nothing}) where {T}
+    value = convert(Union{Some{T},Nothing}, value)
+    set_task_ctxvars(merge_ctxvars(get_task_ctxvars(), (var.key => value,)))
+    return var
+end
+
+"""
+    genkey(__module__::Module, varname::Symbol) -> UUID
+
+Generate a stable UUID for a context variable `__module__.\$varname`.
+"""
+function genkey(__module__::Module, varname::Symbol)
+    pkgid = PkgId(__module__)
+    if pkgid.uuid === nothing
+        throw(ArgumentError(
+            "Module `$__module__` is not a part of a package. " *
+            "`@contextvar` can only be used inside a package.",
+        ))
+    end
+    fullpath = push!(collect(fullname(__module__)), varname)
+    if any(x -> contains(string(x), "."), fullpath)
+        throw(ArgumentError(
+            "Modules and variable names must not contain a dot:\n" * join(fullpath, "\n"),
+        ))
+    end
+    return uuid5(pkgid.uuid, join(fullpath, '.'))
```

Comment:
Ah very nice chaining these UUIDs. I was just coming here to suggest it should be done this way. But you've already done it :-)

Reply:
Good to know that we independently arrived at the same solution :) It's an indication that this solution makes sense.

Reply:
Well perhaps I shouldn't be so definitive as to say "should" ... but I think it makes sense :-)

Reply:
Speaking of UUIDs, where does `uuid5` come from? I guess it will need to be moved it into `base/uuid.jl` from the UUIDs stdlib?

Reply:
There is a fake `uuid5` definition for loading:

https://github.com/JuliaLang/julia/blob/8f512f3f6d5d69403be4b8179bd80ba64229e761/base/loading.jl#L103-L116

Reply:
Ah I see. Not your fault at all, but ick!

It seems some of the hashing stuff should move back into Base (at least some minimal parts of the implementation of SHA1 and RandomDevice, not the whole stdlib). Good hashing and randomness is core functionality which isn't really optional in some of these cases.

Reply:
Here is a PR to move `RandomDevice` to `Base`: https://github.com/JuliaLang/julia/pull/35894